File: 130-surrounded-regions/130-surrounded-regions.cpp
class Solution {
public:
    
    int dx[4] = {0 , 0 , 1 , -1};
int dy[4] = {1 , -1 , 0 , 0};

bool valid(int m , int n , int i , int j){
	if (i >= m || i < 0 || j < 0 || j >= n)
		return false ;
	return true;
}


void dfs(int i , int j , int m , int n , vector<vector<char>>& board){

	board[i][j] = 'T';

	for(int k = 0 ; k < 4 ; k ++){

		int x = dx[k] + i;
		int y = dy[k] + j;

		if(valid(m, n, x, y) && board[x][y] == 'O'){

		dfs(x , y , m , n , board);

		}
	}

}

    
    void solve(vector<vector<char>>& board) {
        	int m = board.size();
	int n = board[0].size();

	for (int i = 0 ; i < m ; i ++){
		for (int j = 0 ; j < n ; j ++){
			if ((i == 0 || i == m - 1 || j == 0 || j == n - 1) && board[i][j] == 'O'){

				dfs(i, j , m , n , board);
			}
		}
	}

	for (int i = 0 ; i < m ; i ++ ){
		for (int j = 0 ; j < n ; j ++ ){
			if (board[i][j] == 'T')
				board[i][j] = 'O';
			else
				board[i][j] = 'X';
		}
	}

    }
};

File: 130-surrounded-regions/130-surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
            rows , cols = len(board) , len(board[0])
            dx = [1,-1,0,0]
            dy = [0,0,1 ,-1]
            def dfs(i , j):
                if i < 0 or i >= rows or j < 0 or j >= cols or board[i][j] != "O":
                    return
                board[i][j] = "T"
                for k in range(4):
                    x = dx[k] + i
                    y = dy[k] + j
                    dfs(x , y)
            for i in range(rows):
                for j in range(cols):
                    if (board[i][j] == "O" and (i in [0 , rows-1] or j in [0 , cols-1])):
                        dfs(i , j)

            for i in range(rows):
                for j in range(cols):
                    if (board[i][j] == "O"):
                        board[i][j] = "X"

            for i in range(rows):
                for j in range(cols):
                    if board[i][j] == "T":
                        board[i][j] = "O"


        